onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py

```diff
@@ -0,0 +1,474 @@
+
+    parser.add_argument(
+        "--quantize_reduce_range",
+        required=False,
+        action="store_true",
+        help="Quantize weights with 7 bits.",
+    )
+    parser.set_defaults(quantize_reduce_range=False)
+
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        action="store_true",
+        help="Print verbose logs",
+    )
+    parser.set_defaults(verbose=False)
+
+    parser.add_argument(
+        "-d",
+        "--use_dynamo_export",
+        action="store_true",
+        help="Use the new Dynamo exporter instead of the old TorchScript exporter",
+    )
+    parser.set_defaults(use_dynamo_export=False)
+
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    args = get_args()
+    setup_logger(args.verbose)
+    prepare_environment(args.input, args.output, args.execution_provider != "cpu")
+    remove_existing_files(args.output)
+    logger.info(f"Arguments: {args}")
+
+    # Load model and config
+    use_auth_token = args.input == os.path.join(".")
+    l_config = LlamaConfig.from_pretrained(
+        args.model_name if use_auth_token else args.input, use_auth_token=use_auth_token
+    )
+    llama = LlamaForCausalLM.from_pretrained(
+        args.model_name if use_auth_token else args.input, use_auth_token=use_auth_token, use_cache=True
+    )
+    original_model_name = args.model_name
+    args.model_name = args.model_name.split("/")[-1]
+
+    # Export to ONNX
+    if args.use_dynamo_export:
+        logger.warning("Please ensure you have installed PyTorch, ONNX, and ONNX Script as follows.")
+        logger.warning("Step 1 - PyTorch nightly: https://pytorch.org/get-started/locally/")
+        logger.warning("Step 2 - ONNX weekly: https://pypi.org/project/onnx-weekly/")
+        logger.warning(
+            "Step 3 - ONNX Script from source: https://github.com/microsoft/onnxscript#installing-onnx-script"
+        )
+        logger.warning(
+            "Note: When you install ONNX weekly, omit `onnx` when running the first line for installing ONNX Script. \
+                        This is because you already installed `onnx-weekly` in the previous step."
+        )
+        run_dynamo_export(args, l_config, llama)
+    else:
+        run_torchscript_export(args, l_config, llama)
+
+    # Change precision of exported models if not FP32
+    decoder_model_fp32_path = os.path.join(args.output, f"{args.model_name}_decoder_model_fp32.onnx")
+    decoder_with_past_model_fp32_path = os.path.join(
+        args.output, f"{args.model_name}_decoder_with_past_model_fp32.onnx"
+    )
+
+    if args.precision == Precision.FLOAT16:
+        # Convert decoder_model.onnx to FP16
+        decoder_model_fp16_path = os.path.join(args.output, f"{args.model_name}_decoder_model_fp16.onnx")
+        model = OnnxModel(onnx.load_model(decoder_model_fp32_path, load_external_data=True))
+        model.convert_float_to_float16(keep_io_types=False, op_block_list=["If"])
```

Comment:
Generally, we need optimize model first before converting to float16.

Reply:
It's fine that we leave it right now.

We can consider converting model to fp16 first, then run graph optimization so that optimization requires less memory.